File: scripts/mcp-jd-app-verify.mjs
/**
 * 京东 App 真机验证（Node.js + ada-mcp MCP stdio）。
 *
 * 用法：
 *   node scripts/mcp-jd-app-verify.mjs
 *   node scripts/mcp-jd-app-verify.mjs --probe
 *   node scripts/mcp-jd-app-verify.mjs --server local
 *
 * 前置（推荐手动常驻，少闪窗）：
 *   set APPIUM_HOME=D:\WORKSPACE\PLAN\ada\APPIUM_HOME
 *   set ANDROID_HOME=<你的 Android SDK>
 *   npx appium --address 127.0.0.1 --port 4723
 */
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { CallToolResultSchema } from "@modelcontextprotocol/sdk/types.js";
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { createAdaMcpTransport, repoRoot } from "./mcp-transport.mjs";

const LONG_REQ = { timeout: 360_000 };
const APP_SESSION = "jd-mcp-verify-app";
const JD_APP_PACKAGE = process.env.ADA_JD_APP_PACKAGE ?? "com.jingdong.app.mall";
const JD_APP_ACTIVITY = process.env.ADA_JD_APP_ACTIVITY ?? ".MainFrameActivity";
const APPIUM_URL = process.env.ADA_APPIUM_URL ?? "http://127.0.0.1:4723";

const WIN_SPAWN = process.platform === "win32" ? { windowsHide: true } : {};

function parseArgs(argv) {
  const args = {
    probe: false,
    server: "local",
    installMobileDeps: false,
    allowMock: false,
    skipAdbLaunch: false,
    appPackage: JD_APP_PACKAGE,
    appActivity: JD_APP_ACTIVITY,
    appiumUrl: APPIUM_URL,
    appWaitMs: 4000,
    outputDir: "artifacts",
    launcherVersion: process.env.ADA_MCP_LAUNCHER_VERSION ?? "0.1.28"
  };
  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === "--probe") args.probe = true;
    else if (a === "--install-mobile-deps") args.installMobileDeps = true;
    else if (a === "--allow-mock") args.allowMock = true;
    else if (a === "--skip-adb-launch") args.skipAdbLaunch = true;
    else if (a === "--server") args.server = argv[++i] ?? "local";
    else if (a === "--app-package") args.appPackage = argv[++i] ?? JD_APP_PACKAGE;
    else if (a === "--app-activity") args.appActivity = argv[++i] ?? JD_APP_ACTIVITY;
    else if (a === "--appium-url") args.appiumUrl = argv[++i] ?? APPIUM_URL;
    else if (a === "--output-dir") args.outputDir = argv[++i] ?? "artifacts";
    else if (a === "--launcher-version") args.launcherVersion = argv[++i] ?? "0.1.28";
    else if (a === "--help" || a === "-h") {
      console.log(`用法: node scripts/mcp-jd-app-verify.mjs [选项]
  --probe                 仅 Appium 探活
  --server local|dev|npm  默认 local（node dist/cli.cjs，少闪 cmd）
  --install-mobile-deps   经 MCP 安装 mobile 依赖（会闪 cmd，仅首次需要）
  --skip-adb-launch       不先用 adb 拉起京东 App
  --app-activity <name>   默认 .MainFrameActivity`);
      process.exit(0);
    }
  }
  return args;
}

function adbSpawn(args, label) {
  const proc = spawnSync("adb", args, {
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
    ...WIN_SPAWN
  });
  if (proc.status !== 0) {
    const err = String(proc.stderr ?? proc.stdout ?? "").trim();
    throw new Error(`${label} 失败: ${err || `exit ${proc.status}`}`);
  }
  return proc;
}

function parseToolPayload(toolResult) {
  if (toolResult?.isError) {
    const text = toolResult.content?.map((b) => b.text).filter(Boolean).join("\n");
    throw new Error(text || "MCP tool error");
  }
  const text = toolResult?.content?.[0]?.text ?? "";
  if (!text) return {};
  try {
    return JSON.parse(text);
  } catch {
    return { raw: text };
  }
}

function assertSuccess(data, step, allowMock) {
  if (data?.success === false) {
    throw new Error(`${step} failed: ${data.errorCode ?? ""} ${data.errorMessage ?? JSON.stringify(data)}`);
  }
  const inner = data?.data;
  if (!allowMock && inner?.mode === "mock" && inner?.reason) {
    throw new Error(`${step} fell back to mock: ${inner.reason}`);
  }
}

async function callTool(client, name, arguments_) {
  const raw = await client.callTool({ name, arguments: arguments_ }, CallToolResultSchema, LONG_REQ);
  return parseToolPayload(raw);
}

function preflightAndroidDevice() {
  let proc;
  try {
    proc = spawnSync("adb", ["devices"], {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "pipe"],
      ...WIN_SPAWN
    });
  } catch {
    return { ok: false, message: "未找到 adb" };
  }
  const lines = String(proc.stdout ?? "")
    .split(/\r?\n/)
    .map((ln) => ln.trim())
    .filter(Boolean);
  const devices = lines.slice(1).filter((ln) => ln.includes("\tdevice"));
  if (devices.length === 0) {
    return { ok: false, message: "无已连接 Android 设备或 USB 未授权" };
  }
  return { ok: true, deviceId: devices[0].split("\t")[0] };
}

function preflightJdAppInstalled(appPackage) {
  const proc = spawnSync("adb", ["shell", "pm", "list", "packages", appPackage], {
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
    ...WIN_SPAWN
  });
  if (!String(proc.stdout ?? "").includes(appPackage)) {
    return { ok: false, message: `未安装 ${appPackage}` };
  }
  return { ok: true };
}

function launchJdAppViaAdb(args, deviceId) {
  const component = `${args.appPackage}/${args.appActivity.startsWith(".") ? args.appActivity : args.appActivity}`;
  console.log(`  [app] adb 启动京东: ${component}`);
  adbSpawn(["-s", deviceId, "shell", "am", "start", "-W", "-n", component], "adb am start");
}

function appPayloadBase(args, deviceId) {
  const activity = args.appActivity.startsWith(".")
    ? `${args.appPackage}${args.appActivity}`
    : args.appActivity;
  const capabilities = {
    platformName: "Android",
    "appium:automationName": "UiAutomator2",
    "appium:deviceName": "Android",
    "appium:udid": deviceId,
    "appium:noReset": true,
    "appium:fullReset": false,
    "appium:autoLaunch": false,
    "appium:appPackage": args.appPackage,
    "appium:appActivity": activity,
    "appium:appWaitActivity": "*",
    "appium:appWaitDuration": 30000,
    "appium:ignoreHiddenApiPolicyError": true,
    "appium:newCommandTimeout": 300
  };
  return {
    real: true,
    serverUrl: args.appiumUrl,
    capabilities
  };
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const outDir = path.isAbsolute(args.outputDir)
    ? args.outputDir
    : path.join(repoRoot, args.outputDir);
  fs.mkdirSync(outDir, { recursive: true });

  const ts = new Date().toISOString().replace(/[-:]/g, "").replace(/\..+/, "").replace("T", "-");
  const shot1 = path.join(outDir, `jd-mcp-app-${ts}-1.png`);
  const shot2 = path.join(outDir, `jd-mcp-app-${ts}-2.png`);
  const steps = [];

  console.log(
    `[app] MCP server=${args.server}  package=${args.appPackage}  activity=${args.appActivity}\n`
  );

  const transport = createAdaMcpTransport({
    server: args.server,
    installDeps: "skip",
    launcherVersion: args.launcherVersion
  });
  const client = new Client({ name: "mcp-jd-app-verify", version: "0.1.0" }, { capabilities: {} });
  await client.connect(transport);

  try {
    console.log("  [app] probe …");
    const probe = await callTool(client, "ada_mobile_action", {
      platform: "android",
      command: "swipe",
      sessionId: APP_SESSION,
      requestId: `jd-app-probe-${ts}`,
      allowMock: true,
      payload: { probe: true }
    });
    steps.push({ name: "probe", ok: true, detail: probe });

    if (args.probe) {
      console.log("  [app] --probe：仅探活\n");
      console.log(JSON.stringify({ ok: true, steps }, null, 2));
      return;
    }

    const dev = preflightAndroidDevice();
    if (!dev.ok) throw new Error(`真机检查: ${dev.message}`);
    console.log(`  [app] 设备: ${dev.deviceId}`);

    const pkg = preflightJdAppInstalled(args.appPackage);
    if (!pkg.ok) throw new Error(pkg.message);

    if (args.installMobileDeps) {
      console.log("  [app] install_mobile_deps（可能闪 cmd）…");
      await callTool(client, "ada_install_deps", { only: "mobile", force: false });
    }

    if (!args.skipAdbLaunch) {
      launchJdAppViaAdb(args, dev.deviceId);
      await sleep(2000);
    }

    const base = appPayloadBase(args, dev.deviceId);

    console.log("  [app] 创建会话并截图 …");
    const open = await callTool(client, "ada_mobile_action", {
      platform: "android",
      command: "screenshot",
      sessionId: APP_SESSION,
      requestId: `jd-app-open-${ts}`,
      allowMock: args.allowMock,
      payload: { ...base, screenshotPath: shot1 }
    });
    assertSuccess(open, "screenshot_home", args.allowMock);
    steps.push({ name: "screenshot_home", ok: true, screenshot: open.data?.screenshot ?? shot1 });
    console.log(`  [app] 截图1: ${open.data?.screenshot ?? shot1}`);

    await sleep(args.appWaitMs);

    async function swipeOnce(index) {
      console.log(`  [app] swipe_right_${index} …`);
      const swipe = await callTool(client, "ada_mobile_action", {
        platform: "android",
        command: "swipe",
        sessionId: APP_SESSION,
        requestId: `jd-app-swipe-${ts}-${index}`,
        allowMock: args.allowMock,
        payload: { ...base, from: [0.2, 0.5], to: [0.8, 0.5] }
      });
      assertSuccess(swipe, `swipe_right_${index}`, args.allowMock);
      steps.push({ name: `swipe_right_${index}`, ok: true });
    }

    for (let i = 0; i < 2; i += 1) {
      try {
        await swipeOnce(i + 1);
      } catch (firstErr) {
        console.warn(`  [app] swipe_right_${i + 1} 失败，等待后重试一次…`);
        await sleep(1500);
        await swipeOnce(i + 1);
      }
      await sleep(500);
    }

    console.log("  [app] 滑动后截图 …");
    const shot = await callTool(client, "ada_mobile_action", {
      platform: "android",
      command: "screenshot",
      sessionId: APP_SESSION,
      requestId: `jd-app-shot2-${ts}`,
      allowMock: args.allowMock,
      payload: { ...base, screenshotPath: shot2 }
    });
    assertSuccess(shot, "screenshot_after_swipe", args.allowMock);
    steps.push({ name: "screenshot_after_swipe", ok: true, screenshot: shot.data?.screenshot ?? shot2 });
    console.log(`  [app] 截图2: ${shot.data?.screenshot ?? shot2}`);

    console.log("\n[app] 完成\n");
    console.log(JSON.stringify({ ok: true, steps }, null, 2));
  } finally {
    await client.close();
  }
}

main().catch((error) => {
  console.error(
    JSON.stringify(
      { ok: false, message: error instanceof Error ? error.message : String(error) },
      null,
      2
    )
  );
  console.error(`
[排查]
  1) 先手动启动 Appium（推荐，避免脚本反复拉起）:
     set APPIUM_HOME=${path.join(repoRoot, "APPIUM_HOME")}
     npx appium --address 127.0.0.1 --port 4723
  2) adb devices 显示 device
  3) npx appium driver install uiautomator2
  4) 使用: node scripts/mcp-jd-app-verify.mjs --server local
  5) 勿用 python --real-app（已废弃 App 路径）`);
  process.exit(1);
});

File: scripts/mcp_jd_app_verify.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
京东 App 真机验证（Python + ada-mcp MCP stdio）。

与 Node 版等价：
  node scripts/mcp-jd-app-verify.mjs --server local

依赖：
  pip install -r scripts/requirements-mcp-verify.txt

用法：
  python scripts/mcp_jd_app_verify.py
  python scripts/mcp_jd_app_verify.py --probe
  python scripts/mcp_jd_app_verify.py --server local

前置（推荐手动常驻 Appium，少闪窗）：
  set APPIUM_HOME=D:\\WORKSPACE\\PLAN\\ada\\APPIUM_HOME
  npx appium --address 127.0.0.1 --port 4723
"""

from __future__ import annotations

import argparse
import asyncio
import json
import os
import subprocess
import sys
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

APP_SESSION = "jd-mcp-verify-app"
JD_APP_PACKAGE = "com.jingdong.app.mall"
JD_APP_ACTIVITY = ".MainFrameActivity"
DEFAULT_APPIUM_URL = "http://127.0.0.1:4723"
TIMEOUT_SEC = 360.0

_WIN_CREATE_NO_WINDOW = getattr(subprocess, "CREATE_NO_WINDOW", 0)


@dataclass
class StepResult:
    name: str
    ok: bool
    detail: dict[str, Any] = field(default_factory=dict)
    error: str | None = None


def repo_root() -> Path:
    return Path(__file__).resolve().parents[1]


def _subprocess_kwargs() -> dict[str, Any]:
    kw: dict[str, Any] = {}
    if sys.platform == "win32" and _WIN_CREATE_NO_WINDOW:
        kw["creationflags"] = _WIN_CREATE_NO_WINDOW
    return kw


def adb_run(args: list[str], label: str) -> subprocess.CompletedProcess[str]:
    proc = subprocess.run(
        ["adb", *args],
        capture_output=True,
        text=True,
        timeout=60,
        check=False,
        **_subprocess_kwargs(),
    )
    if proc.returncode != 0:
        err = (proc.stderr or proc.stdout or "").strip()
        raise RuntimeError(f"{label} 失败: {err or proc.returncode}")
    return proc


def parse_tool_payload(result: Any) -> dict[str, Any]:
    if getattr(result, "isError", False):
        parts = [getattr(b, "text", "") for b in result.content or [] if getattr(b, "text", None)]
        raise RuntimeError("MCP tool error: " + ("\n".join(parts) or "unknown"))
    texts = [getattr(b, "text", "") for b in result.content or [] if getattr(b, "text", None)]
    raw = "\n".join(texts).strip()
    if not raw:
        return {}
    try:
        parsed = json.loads(raw)
        return parsed if isinstance(parsed, dict) else {"value": parsed}
    except json.JSONDecodeError:
        return {"raw": raw}


def assert_success(data: dict[str, Any], step: str, *, allow_mock: bool = False) -> None:
    if data.get("success") is False:
        raise RuntimeError(
            f"{step} failed: {data.get('errorCode') or ''} {data.get('errorMessage') or data}"
        )
    inner = data.get("data") if isinstance(data.get("data"), dict) else {}
    if not allow_mock and inner.get("mode") == "mock" and inner.get("reason"):
        raise RuntimeError(f"{step} fell back to mock: {inner.get('reason')}")


async def call_tool(
    session: ClientSession,
    name: str,
    arguments: dict[str, Any],
) -> dict[str, Any]:
    result = await asyncio.wait_for(session.call_tool(name, arguments), timeout=TIMEOUT_SEC)
    return parse_tool_payload(result)


def apply_local_mobile_env(env: dict[str, str], root: Path) -> None:
    android_home = os.environ.get("ANDROID_HOME", "").strip()
    if android_home and Path(android_home).is_dir():
        env["ANDROID_HOME"] = android_home
        env.setdefault("ANDROID_SDK_ROOT", android_home)
    for name in ("APPIUM_HOME", "ANDROID_HOME", "ANDROID_SDK_ROOT"):
        candidate = root / name
        if candidate.is_dir():
            env[name] = str(candidate)
            if name == "ANDROID_HOME":
                env["ANDROID_SDK_ROOT"] = str(candidate)


def build_server_params(server: str, root: Path, launcher_version: str) -> StdioServerParameters:
    env = {**os.environ}
    env["ADA_MCP_INSTALL_DEPS"] = "skip"
    env["ADA_MCP_SKIP_INSTALL_DEPS"] = "1"
    apply_local_mobile_env(env, root)

    if server == "dev":
        cli_ts = root / "apps" / "ada-mcp-server" / "src" / "cli.ts"
        tsx = root / "node_modules" / "tsx" / "dist" / "cli.mjs"
        if not cli_ts.is_file():
            raise FileNotFoundError(f"dev 入口不存在: {cli_ts}")
        if not tsx.is_file():
            raise FileNotFoundError(f"未找到 tsx: {tsx}\n请先 npm install")
        env["ADA_MCP_SERVER_ENTRY"] = str(cli_ts)
        return StdioServerParameters(
            command=os.environ.get("ADA_NODE", "node"),
            args=[str(tsx), str(cli_ts)],
            cwd=str(root),
            env=env,
        )

    if server == "local":
        cli_cjs = root / "apps" / "ada-mcp-server" / "dist" / "cli.cjs"
        if not cli_cjs.is_file():
            raise FileNotFoundError(
                f"本地 bundle 不存在: {cli_cjs}\n请先: npm run build:npm -w @ada-mcp/mcp-server"
            )
        env["ADA_MCP_SERVER_ENTRY"] = str(cli_cjs)
        return StdioServerParameters(
            command=os.environ.get("ADA_NODE", "node"),
            args=[str(cli_cjs)],
            cwd=str(root),
            env=env,
        )

    pnpm = "pnpm.cmd" if sys.platform == "win32" else "pnpm"
    spec = f"@ada-mcp/launcher@{launcher_version}"
    return StdioServerParameters(
        command=pnpm,
        args=["dlx", spec],
        cwd=str(root),
        env=env,
    )


def preflight_android_device() -> tuple[bool, str]:
    try:
        proc = subprocess.run(
            ["adb", "devices"],
            capture_output=True,
            text=True,
            timeout=15,
            check=False,
            **_subprocess_kwargs(),
        )
    except FileNotFoundError:
        return False, "未找到 adb"
    lines = [ln.strip() for ln in proc.stdout.splitlines() if ln.strip()]
    devices = [ln for ln in lines[1:] if "\tdevice" in ln]
    if not devices:
        unauthorized = [ln for ln in lines[1:] if "\tunauthorized" in ln]
        if unauthorized:
            return False, "设备未授权 USB 调试"
        return False, "无已连接 Android 设备"
    return True, devices[0].split("\t")[0]


def preflight_jd_app_installed(app_package: str) -> None:
    proc = subprocess.run(
        ["adb", "shell", "pm", "list", "packages", app_package],
        capture_output=True,
        text=True,
        timeout=30,
        check=False,
        **_subprocess_kwargs(),
    )
    if app_package not in (proc.stdout or ""):
        raise RuntimeError(f"未安装 {app_package}")


def launch_jd_app_via_adb(app_package: str, app_activity: str, device_id: str) -> None:
    activity = app_activity if not app_activity.startswith(".") else app_activity
    component = f"{app_package}/{activity}"
    print(f"  [app] adb 启动京东: {component}")
    adb_run(["-s", device_id, "shell", "am", "start", "-W", "-n", component], "adb am start")


def app_payload_base(
    *,
    app_package: str,
    app_activity: str,
    appium_url: str,
    device_id: str,
) -> dict[str, Any]:
    activity = (
        f"{app_package}{app_activity}" if app_activity.startswith(".") else app_activity
    )
    return {
        "real": True,
        "serverUrl": appium_url,
        "capabilities": {
            "platformName": "Android",
            "appium:automationName": "UiAutomator2",
            "appium:deviceName": "Android",
            "appium:udid": device_id,
            "appium:noReset": True,
            "appium:fullReset": False,
            "appium:autoLaunch": False,
            "appium:appPackage": app_package,
            "appium:appActivity": activity,
            "appium:appWaitActivity": "*",
            "appium:appWaitDuration": 30000,
            "appium:ignoreHiddenApiPolicyError": True,
            "appium:newCommandTimeout": 300,
        },
    }


async def swipe_once(
    session: ClientSession,
    *,
    ts: str,
    index: int,
    base: dict[str, Any],
    allow_mock: bool,
    steps: list[StepResult],
) -> None:
    print(f"  [app] swipe_right_{index} …")
    data = await call_tool(
        session,
        "ada_mobile_action",
        {
            "platform": "android",
            "command": "swipe",
            "sessionId": APP_SESSION,
            "requestId": f"jd-app-swipe-{ts}-{index}",
            "allowMock": allow_mock,
            "payload": {**base, "from": [0.2, 0.5], "to": [0.8, 0.5]},
        },
    )
    assert_success(data, f"swipe_right_{index}", allow_mock=allow_mock)
    steps.append(StepResult(name=f"swipe_right_{index}", ok=True))


async def run_verify(args: argparse.Namespace) -> list[StepResult]:
    root = repo_root()
    out_dir = Path(args.output_dir)
    if not out_dir.is_absolute():
        out_dir = root / out_dir
    out_dir.mkdir(parents=True, exist_ok=True)

    ts = datetime.now().strftime("%Y%m%d-%H%M%S")
    shot1 = out_dir / f"jd-mcp-app-{ts}-1.png"
    shot2 = out_dir / f"jd-mcp-app-{ts}-2.png"
    steps: list[StepResult] = []

    print(
        f"[app] MCP server={args.server}  package={args.app_package}  "
        f"activity={args.app_activity}\n"
    )

    params = build_server_params(args.server, root, args.launcher_version)

    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            await asyncio.wait_for(session.initialize(), timeout=120)

            print("  [app] probe …")
            probe = await call_tool(
                session,
                "ada_mobile_action",
                {
                    "platform": "android",
                    "command": "swipe",
                    "sessionId": APP_SESSION,
                    "requestId": f"jd-app-probe-{ts}",
                    "allowMock": True,
                    "payload": {"probe": True},
                },
            )
            steps.append(StepResult(name="probe", ok=True, detail=probe))

            if args.probe:
                print("  [app] --probe：仅探活\n")
                return steps

            ok, dev_msg = preflight_android_device()
            if not ok:
                raise RuntimeError(f"真机检查: {dev_msg}")
            print(f"  [app] 设备: {dev_msg}")

            preflight_jd_app_installed(args.app_package)

            if args.install_mobile_deps:
                print("  [app] install_mobile_deps（可能闪 cmd）…")
                await call_tool(
                    session,
                    "ada_install_deps",
                    {"only": "mobile", "force": False},
                )

            if not args.skip_adb_launch:
                launch_jd_app_via_adb(args.app_package, args.app_activity, dev_msg)
                await asyncio.sleep(2)

            base = app_payload_base(
                app_package=args.app_package,
                app_activity=args.app_activity,
                appium_url=args.appium_url,
                device_id=dev_msg,
            )

            print("  [app] 创建会话并截图 …")
            open_shot = await call_tool(
                session,
                "ada_mobile_action",
                {
                    "platform": "android",
                    "command": "screenshot",
                    "sessionId": APP_SESSION,
                    "requestId": f"jd-app-open-{ts}",
                    "allowMock": args.allow_mock,
                    "payload": {**base, "screenshotPath": str(shot1)},
                },
            )
            assert_success(open_shot, "screenshot_home", allow_mock=args.allow_mock)
            path1 = (open_shot.get("data") or {}).get("screenshot") or str(shot1)
            steps.append(StepResult(name="screenshot_home", ok=True, detail={"screenshot": path1}))
            print(f"  [app] 截图1: {path1}")

            await asyncio.sleep(args.app_wait_ms / 1000)

            for i in range(2):
                idx = i + 1
                try:
                    await swipe_once(
                        session,
                        ts=ts,
                        index=idx,
                        base=base,
                        allow_mock=args.allow_mock,
                        steps=steps,
                    )
                except Exception:
                    print(f"  [app] swipe_right_{idx} 失败，等待后重试一次…", file=sys.stderr)
                    await asyncio.sleep(1.5)
                    await swipe_once(
                        session,
                        ts=ts,
                        index=idx,
                        base=base,
                        allow_mock=args.allow_mock,
                        steps=steps,
                    )
                await asyncio.sleep(0.5)

            print("  [app] 滑动后截图 …")
            shot2_data = await call_tool(
                session,
                "ada_mobile_action",
                {
                    "platform": "android",
                    "command": "screenshot",
                    "sessionId": APP_SESSION,
                    "requestId": f"jd-app-shot2-{ts}",
                    "allowMock": args.allow_mock,
                    "payload": {**base, "screenshotPath": str(shot2)},
                },
            )
            assert_success(shot2_data, "screenshot_after_swipe", allow_mock=args.allow_mock)
            path2 = (shot2_data.get("data") or {}).get("screenshot") or str(shot2)
            steps.append(
                StepResult(name="screenshot_after_swipe", ok=True, detail={"screenshot": path2})
            )
            print(f"  [app] 截图2: {path2}")

    return steps


def main() -> int:
    parser = argparse.ArgumentParser(
        description="京东 App 真机 ada-mcp 验证（与 mcp-jd-app-verify.mjs 等价）",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )
    parser.add_argument("--probe", action="store_true", help="仅 Appium 探活")
    parser.add_argument(
        "--server",
        choices=("local", "dev", "npm"),
        default="local",
        help="MCP 入口，默认 local",
    )
    parser.add_argument(
        "--install-mobile-deps",
        action="store_true",
        help="经 MCP 安装 mobile 依赖（仅首次）",
    )
    parser.add_argument("--allow-mock", action="store_true")
    parser.add_argument("--skip-adb-launch", action="store_true", help="不先用 adb 启京东")
    parser.add_argument(
        "--app-package",
        default=os.environ.get("ADA_JD_APP_PACKAGE", JD_APP_PACKAGE),
    )
    parser.add_argument(
        "--app-activity",
        default=os.environ.get("ADA_JD_APP_ACTIVITY", JD_APP_ACTIVITY),
    )
    parser.add_argument(
        "--appium-url",
        default=os.environ.get("ADA_APPIUM_URL", DEFAULT_APPIUM_URL),
    )
    parser.add_argument("--app-wait-ms", type=int, default=4000)
    parser.add_argument("--output-dir", default="artifacts")
    parser.add_argument(
        "--launcher-version",
        default=os.environ.get("ADA_MCP_LAUNCHER_VERSION", "0.1.27"),
    )
    args = parser.parse_args()

    try:
        steps = asyncio.run(run_verify(args))
        ok = all(s.ok for s in steps)
        print("\n[app] 完成\n")
        print(json.dumps({"ok": ok, "steps": [{"name": s.name, "ok": s.ok} for s in steps]}, ensure_ascii=False))
        return 0 if ok else 1
    except Exception as exc:
        print(json.dumps({"ok": False, "message": str(exc)}, ensure_ascii=False), file=sys.stderr)
        root = repo_root()
        print(
            f"""
[排查]
  1) 手动启动 Appium:
     set APPIUM_HOME={root / "APPIUM_HOME"}
     npx appium --address 127.0.0.1 --port 4723
  2) adb devices 显示 device
  3) python scripts/mcp_jd_app_verify.py --server local
  4) Node 等价: node scripts/mcp-jd-app-verify.mjs --server local""",
            file=sys.stderr,
        )
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
